crates/iota-types/src/scoring_metrics.rs

```diff
@@ -0,0 +1,294 @@
+                    metric.store(0, Ordering::Relaxed);
+                }
+                for metric in &metrics.equivocations {
+                    metric.store(0, Ordering::Relaxed);
+                }
+                for metric in &metrics.missing_proposals {
+                    metric.store(0, Ordering::Relaxed);
+                }
+            }
+        }
+    }
+}
+
+impl VersionedScoringMetrics {
+    // Given a VersionedMisbehaviorReport received from another authority, we use
+    // this method to update the received scoring metrics counts. To avoid
+    // updates to be dependent on the order they are applied, we only effectively
+    // update counts that are increased by the report. This also means that any type
+    // of metric contained in this struct must be guaranteed to be monotonically
+    // increasing. Example: number of faulty blocks detected for a given authority
+    // is monotonically increasing by design; average faulty blocks per minute is
+    // not.
+    pub fn update_from_report(&self, report: &VersionedMisbehaviorReport) {
+        match (self, report) {
+            (VersionedScoringMetrics::V1(metrics), VersionedMisbehaviorReport::V1(report_v1)) => {
+                for (i, value) in report_v1.faulty_blocks_provable.iter().enumerate() {
+                    if *value > metrics.faulty_blocks_provable[i].load(Ordering::Relaxed) {
+                        metrics.faulty_blocks_provable[i].store(*value, Ordering::Relaxed);
+                    }
+                }
+                for (i, value) in report_v1.faulty_blocks_unprovable.iter().enumerate() {
+                    if *value > metrics.faulty_blocks_unprovable[i].load(Ordering::Relaxed) {
+                        metrics.faulty_blocks_unprovable[i].store(*value, Ordering::Relaxed);
+                    }
+                }
+                for (i, value) in report_v1.equivocations.iter().enumerate() {
+                    if *value > metrics.equivocations[i].load(Ordering::Relaxed) {
+                        metrics.equivocations[i].store(*value, Ordering::Relaxed);
+                    }
+                }
+                for (i, value) in report_v1.missing_proposals.iter().enumerate() {
+                    if *value > metrics.missing_proposals[i].load(Ordering::Relaxed) {
+                        metrics.missing_proposals[i].store(*value, Ordering::Relaxed);
+                    }
+                }
+            }
+        }
+    }
+
+    // Given a VersionedMisbehaviorReport, create a VersionedScoringMetrics struct
+    // with the same values. Used when an authority receives a report from the
+    // network and needs to create a local copy of the metrics contained in it.
+    pub fn from_report(report: &VersionedMisbehaviorReport) -> Self {
+        match report {
+            VersionedMisbehaviorReport::V1(report_v1) => {
+                let committee_size = report_v1.faulty_blocks_provable.len();
```

Comment:
is the report validated anywhere that it the number of elements matches the committee size?

Reply:
added to https://github.com/iotaledger/iota/pull/9175